### ISSUE TITLE
Allow internationalization of Twinkle

### ISSUE THREAD

**Issue by Huji** [azatoth/twinkle#129]
I would like to suggest Twinkle to be internationalized. This will allow easier translation and updating of Twinle on non-English projects.

All we need at this point, is to store the messages in a separate file (or even in the same file, but in a separate array).

**Comment by Huji** [comment]
More than 2.5 years have passed and this issue is still unresolved. Moreover, Twinkle has been modularized since then, and Gadgets 2.0 has not been released on Wikipedia. I think it is time to get serious with this.

I would like to take this bug myself. Here is what I plan to do:

1. Add a new module to Twinkle, called twinklelang.js in which I will create an object that contains the messages for each language, as well as a set of methods for that object that can retrieve the messages (or fallback to English when no translation is found).

2. Replace all hardcoded messages throughout Twinkle with references to the messages defined in twinklelang.js (although for the time being, I will NOT make console log messages internationalized; only content that is shown in the browser window (messages, link texts, etc) will be internationalized.

Specific suggestions about the above are welcome. If someone assigns this issue to me that would be great too (although you don't have to).

**Comment by mc10** [comment]
If you're willing to take this on, then definitely props to you, but be warned that this will require quite a bit of effort. Style-wise I'm okay with a WordPress-style `__()` (double underscore) function to wrap around all interface strings, and let that function determine which string to display.

**Comment by atlight** [comment]
Okay, a couple of suggestions:

* Start with a simple module first (e.g. twinkleshared) and show your work.
* One of the real challenges is that, to localise Twinkle, it is not only the message strings that need to be translated. The actual templates and options themselves (for example, the welcome templates, shared IP templates, list of CSD criteria, etc), with their attendant subgroups, parameters, etc. need to be localised. That is why this work hasn't been completed yet - it is a truly enormous undertaking that essentially amounts to "rewriting Twinkle".

If you are going to do this, I would suggest you have a look at the prior attempts at this: Siddhartha Ghai's "TWG" or "TW Global" (probably stored on enwiki somewhere, you may have to hunt for this one) and my initial work on this (now in the branch "twinklelocal" of this repository). It's a big challenge and it will be interesting to see your approach.

If you do end up "rewriting Twinkle", it would be nice to take the opportunity to ditch XML and switch to JSON :)

**Comment by Huji** [comment]
Thanks for both comments!

@mc10 I am aware of how much work it is going to be; that is the primary reason I hesitate to get to this for over two years. I have some free time at hand now, so might as well get this done!

@atlight That is exactly how I wanted to approach this: by starting from a small module and validation my approach with you and others who have more experience with Twinkle. I have seen TWG and I am thinking of a similar approach, except I was thinking of using either JSON or just plain JS objects. Something like:

````
Twinkle.lang = {
  Twinkle.lang.parse = function(msgid, params = false) {
      // params can be a string or an array of strings, and is used to parse the message
      ....
  }

  Twinkle.lang.messages = {
    en: {
      msg-portlet-name : 'TW',
      msg-confirm-delete : 'Are you sure you want to delete $1?'
      ....
    },
    de: {
      ...
     }
  }
};
````

And then you would call it like:

````
/* Example from Twinkle.js */
...
Twinkle.defaultConfig.twinkle.portletName = Twinkle.lang.parse('msg-portlet-name');
somethingElse = Twinkle.lang.parse('msg-confirm-delete', pageName);
````

**Comment by mc10** [comment]
`Twinkle.lang.parse` is probably too long&mdash;you don't want to clutter up the code. In this case I think a short global function is okay. Also, if we're actually going to do this all the way, separate files for each language would probably be for the best.

The structure looks good though!

**Comment by Huji** [comment]
@mc10 what do you recommend for the function name? Twinkle.__() ?

Also, as far as the messages, one alternative is to have every module contain its own messages (at the end). Another is to create separate JS file for each module (so then each module becomes two file, one with code, one with messages). I vote for the first, but I appreciate feedback.

**Comment by atlight** [comment]
In my "twinklelocal" branch I used the approach of aliasing a function by passing it into the global wrapper function for the module. Check it out.

I would suggest to keep the modules in separate files, as they are now, and move the localisation material into its own file. This means that, when an administrator is localising Twinkle to their own wiki, they will only need to modify the one file.

**Comment by atlight** [comment]
PS: We have an IRC channel `#wikipedia-userscripts` which you can use. I am sometimes online there.

**Comment by Huji** [comment]
On my way to that IRC channel. A link to your "twinklelocal" branch would be helpful as well.

**Comment by Huji** [comment]
Please see https://github.com/Huji/twinkle/commit/be1a412410f8f2ae3fd45771721d099a7a75bb6e

The locale is set only once (and in reality, it wouldn't be hard-coded, but would rather be fetched from the wiki's config.

I am using jquery.i18n which is a module created for Wikimedia specifically, by James Forrest; it is now part of ULS (which all Wikipedia's use) and very soon will be part of core. You need to add it as a dependency like [this](https://fawp.wmflabs.org/w/index.php?title=MediaWiki:Gadgets-definition&diff=prev&oldid=130).

Because I envision jquery.i18n to be used by other Gadgets too, I am prefixing all Twinkle messages with `tw-`

Do you like this style? Before I invest more time in it, please let me know if there is something that needs to be changed. My plan is to finish the entire i18nization in about one week. Once merged into core, then I will start working on adding l10n too (localization will be useful when different wikis use different criteria for speedy deletion, etc. I am keeping l10n separate from i18n for good reasons).

**Comment by kaldari** [comment]
I would strongly suggest using JSON for the i18n data. This will open up possibilities like using the JsonConfig extension to provide a nice UI for editing the messages on-wiki and possibly porting Twinkle into a MediaWiki extension one day (all MediaWIki extensions use JSON files for storing i18n data, separated by language, e.g. en.json, de.json, etc.).

**Comment by Huji** [comment]
This is part of the reason I have been stalling on this project. I want to solicit more feedback. We can't have it both ways: we have to either split the files by language (en.json, de.json, ...), or by module (one for Twinkleprod, one for TwinkleCSD, ...). I am personally in favor of the former, and atlight seems to like the latter. We need to make this decision now, before I create the files.

**Comment by Hello71** [comment]
@Huji split by language is better. see: gettext. this way makes much more sense especially for more languages; each person needs to edit their own set of languages, instead of all the modules then searching for their languages.

**Comment by Huji** [comment]
Sounds good. In that case, I will make it happen in a few days (in my own fork of course).
One important thing I noticed is this: i18n should be called BEFORE any values are set. Right now, twinkle sets certain config variables, including the one that determines the portlet's label ("TW") before anything else is loaded. I am going to change it such that i18n is loaded before anything else.

**Comment by atlight** [comment]
Please see also the discussion re this issue at https://phabricator.wikimedia.org/T108426. If you think you will have time to complete work on this issue within the next few months, please comment there.

There is actually very little pure i18n in Twinkle. Most is actually localisation (depends on the individual wiki, not just the language). For example, just about every string in every module should be customizable by the wiki, perhaps with the exception of some standard error messages. 

What I actually envisage is a separate file for each language-module pair... is that a bit much?

**Comment by Huji** [comment]
@atlight I think it is an overkill to have one file for each language-module pair. Instead, I think I will name the messages in such a way that you can easily determine which module they belong to (e.g. "tw-csd-...") but keep them all in the same file for each language.

As for i18n vs l10n, your point is well taken. Different modules may not apply to certain wikis (e.g. not every wiki has a PROD policy), and certain module features may need to be modified or deactivated (e.g. not all tags in the Tags module have an equivalent in every other wiki, and there are other tags that certain wikis have that En WP doesn't).

Nonetheless, the overlaps is so large that I think simply enabling i18n can be a huge boost in enabling the use of TW in many wikis. That is my humble opinion, and I hope this discussion will lead us to a consensus; preferably before I start spending hours on i18n.

Lastly, I will definitely check out T198426 and comment there if I can help.

**Comment by atlight** [comment]
The only reason I was against organising the i18n/l10n data into files by language is because I don't expect other wikis' config data to be kept in this repository. I certainly don't want to have to manage it. My vision is to have only enwiki's config data here, and other wikis can keep theirs elsewhere. And as I said above, most i18n is actually wiki-specific, so I'm not sure that it makes sense to store even that in this repository.

But let's not worry about that for now. Start coding the way you see fit, and we can discuss it further once you have made a start. I'm keen to see how your first module turns out. 

Just a tip - there's a reason my initial "twinklelocal" effort started off with the Shared IP module - it's the shortest of the lot (apart from "diff", which isn't very interesting). It's probably the best one for getting your infrastructure up and running.

**Comment by Huji** [comment]
All right I finally decided to start with jquery.i18n for now so we have something to discuss. Please see https://github.com/Huji/twinkle/commit/6982ff08b17d161231019fbbc217c8595804b4d4

Note that any initialization that involves interface messages must be done after i18n is initialized therefore I had to move some parts of the code into initialize() functions and call them when the module is loaded.

Also note that not all messages are i18n'ed in this commit. There are two reasons. The less important one is I didn't have the time. The more important one is if a message contains the string '{{', jquery.i18n tries to parse what comes after it and that is a problem with the SharedIP module of twinkle, because it has strings that start with {{.

I have posted an issue on jquery.i18n asking for a solution. Unless they provide a solution quickly, we may be better off creating our own i18n module.

To test, please check this out: https://fawp.wmflabs.org/wiki/User_talk:1.2.3.4 (after you create an account and enable Twinkle in Special:Preferences of course!) and also https://fawp.wmflabs.org/wiki/User_talk:1.2.3.4?uselang=fa (where you will see that the portlet's label is not TW anymore and changes to "توینکل" as expected).

Comments are appreciated.

**Comment by atlight** [comment]
This is a good start, Huji. 

There are two main issues I have with this, such that I would not be prepared to merge any code that uses this structure:

Firstly, you're missing the localisation aspect. What you are currently doing would be good if, say, we wanted to allow Persian speakers to use Twinkle on English Wikipedia. But that is not a goal of the project, as I see it; instead, we want to make Twinkle usable in the context of other wikis, which this does not achieve.

For example, instead of localising the template names and labels of the individual shared IP templates, you need to provide a way for each wiki to substitute its own list of shared IP templates. For example, imagine a hypothetical project "English Wikicooking". It has three shared IP tags: {{shared IP}}, {{shared IP restaurant}}, {{shared IP retail}}. The {{shared IP}} template supports only the "organisation" parameter, and the other two support only the "contact" parameter. With your proposed patch, there is no way for English Wikicooking to implement the shared IP module to suit their wiki's needs.

Secondly, it's not clear why you are moving stuff around in `twinkle.js`. The whole thing gets wrapped with an anonymous function by ResourceLoader in the end. I suggest you leave the config settings stuff where it is, and simply override the default (English) values as required in your initialisation function.

And yes, if there's no way to escape {{ in jQuery localisation strings, that's a big problem. Can you link to the issue?

**Comment by Huji** [comment]
As for the first issue: I think the best approach to enable both i18n and l10n is to make sure all our modules start with an initialization phase where the "default" values are loaded. So if the templates used in Shared IP for a different language (or wiki) are different, you can override the default config locally. I will try to make an example of it on my fork.
As for the second: the reason I moved stuff around was because my philosophy was that nothing, even the default values, should not set an interface message *before* i18n is loaded. You can do it the other way too: load i18n afterwards, and then override the default configs. I think it is redundant but if you think that is better coding, I will do so.

**Comment by atlight** [comment]
I guess my concern with the second point is, while your approach is probably necessary for non-ResourceLoader scenarios, it is probably not necessary in RL (real life ResourceLoader) scenarios. (I say "probably" because JavaScript loading can a very fickle thing!) I don't really mind that much, though. 

The first issue is more of a concern. My idea was to get rid of the enwiki-specific template stuff from the module entirely, so that the module is mostly wiki-agnostic. (Some of the logic may still have to remain wiki-specific, but that shouldn't be too great an issue in Shared IP.) 

I'd at least like to see an attempt at implementing shared IP for another wiki (I don't know if fawiki has shared IP templates; if not, at least just implement it for my Wikicooking example, or another real wiki that does have shared IP templates.) I'll be interested to see the example in your fork.

**Comment by Huji** [comment]
We have to decide which parts can be wiki-specific and which parts cannot. Take SharedIP as an example; do we only want to allow the templates to be wiki specific (my thought), or both the templates and the textboxes (e.g. one wiki might decide they need a fourth textbox in addition to owner, host and contact info)? Do we want the logic of disabling textboxes to be customizable? (right now if you pick *Shared IP edu* the third textbox gets enabled) And if so, how?

The more flexible we make it, the more complex it becomes but then you can have Twinkle on Wikimedia and non-Wikimedia wikis (as long as you are willing to configure it with all its complexities). If we keep it very simple then we will be limited, perhaps only to Wikipedias, but it would be much easier to adopt.

Right now, my thought is to allow only customizing the template list and forgo the entire logic of disabling textboxes (as it changes from wiki to wiki). One might not like the idea of removing the logic that is already there, but we have to realize that what we like is not important; having a tool that is easily usable to a larger group is important.

TL;DR: Twinkle is highly en-wiki biased. If we are wiling to make it i18n'ed and l10n'ed, we have to swallow the bitter pill of losing some unnecessary en-wiki specific features (like the textbox that is disabled).

**Comment by Huji** [comment]
@atlight here is another step forward. Check my fork again please and test on https://fawp.wmflabs.org/w/index.php?title=User_talk:1.2.3.4&action=edit&redlink=1 and https://fawp.wmflabs.org/w/index.php?title=User_talk:1.2.3.4&action=edit&redlink=1&uselang=de

As you can see, I am not putting the default settings defined in Twinkle.js in a function any more. I simply override Twinkle.defaultConfig.twinkle.portletName further down.

Also, as you can see, as long as the config of each module is in a property which can be overriden before the module is loaded, one can easily localize the modules. In this case, the DE version will have a much shorter list of templates for SharedIP.

All the stuff that is currently added in the middle of Twinkle.js will eventually move into a module called Gadget-twinklelocalize.js for a cleaner look

Also, I switched from jquery.i18n to jquery-i18n (dash versus dot) because it didn't have the issues with "{{". I further customized jquery-18n. The original version is on https://github.com/recurser/jquery-i18n

**Comment by Huji** [comment]
Finally got a chance to clean up my code. Please see https://github.com/azatoth/twinkle/pull/292 and use the following links for a demo:

https://fawp.wmflabs.org/w/index.php
https://fawp.wmflabs.org/w/index.php?uselang=de
https://fawp.wmflabs.org/w/index.php?title=User_talk:1.2.3.4&action=edit&redlink=1 
https://fawp.wmflabs.org/w/index.php?title=User_talk:1.2.3.4&action=edit&redlink=1&uselang=de

With the latter two links, make sure you bring up the Shared IP dialog to notice the difference in content (localized).

**Comment by Huji** [comment]
At this point we need to decide if we want to keep the EN WP-specific features in the modules as *default* values, or if we want to even move those into the localization module. I vote for keeping them where they are, but comments are certainly welcome.

We should also decide if you just want me to cleanup the code in #292 so it is actually pulled into the repo, or if you rather I spend time i18nizing every module first. I am undecided here, and regardless of which we decide I would appreciate help in the i18nization of modules.

**Comment by Huji** [comment]
I created a new pull request showing where I am right now (there was a huge hiatus in my work on this):

https://github.com/azatoth/twinkle/pull/320

It moves l10n and i18n to separate files, keeping all messages in the same place. We can decide if we only want to include the english messages, or that for all languages (we can use Translatewiki for translations if need be). I vote for EN only. For now, I included some DE sample so you can see how things will look like.

Comments are appreciated.

**Comment by Huji** [comment]
I have explained my concerns as comments on the pull request above, but I want to summarize them here:

We all agree it is best if we don't have ANY project-specific (e.g. ENWP-specific) logic in the Twinkle code. The logic should be in the localization file. But to change the entire Twinkle in that way is really hard and takes lots of time.

My approach is: let's first not move any of the logic, but just take a baby step: move all messages out and make them easier to translate. Once that is done, we will move the callbacks in the next step. Is that reasonable?

**Comment by atlight** [comment]
I largely agree with your approach, except it misses some detail. 

Twinkle localisation will involve (from simplest to most complicated):
* Project-independent strings ("Twinkle", "Submit Query", "Reason:", "Notify page creator if possible", etc)
* Project-dependent strings ("Criteria for speedy deletion", "Semi-protection", etc)
* Miscellaneous project-dependent parameters (whether to enable the PROD module on this project)
* Project-dependent objects (lists of templates, etc)
* Project-dependent logic (callbacks, etc)

Which of those levels do you want to attempt? I imagine you'd be interested in the first two, and possibly the next two, but not the last one for now.

**Comment by Huji** [comment]
That is correct. I think it would be logical to do 1, 2 and 4 before we
tackle 3 and 5.

**Comment by Huji** [comment]
Can we please revisit this? I have a pull request that I want to build upon, so long as there is interest in this.

**Comment by Huji** [comment]
See relevant discussion here: https://phabricator.wikimedia.org/T110148

I refrain from working on this until I see actual support (like "do this") rather than mere criticism (like "don't do that") and a true support for having a partial solution in production (as opposed to expecting a complete solution before pushing anything into production)

**Comment by DannyS712** [comment]
@Amorymeltzer I might be interested in working on this, in part because it may make #389 easier. Is this still a pertinent request?

**Comment by Amorymeltzer** [comment]
@DannyS712 Absolutely, and you are welcome to, but I guarantee you're underestimating the amount of work it will require.

**Comment by Huji** [comment]
I hate to be a naysayer, but I agree with @Amorymeltzer that it takes a lot of work and, more importantly, I think this task requires some level of "political will" so to speak, from the owners of the Twinkle code-base that, at least at the time, was non-existent.

The best strategy, in my humble opinion, is to break this down into smaller tasks, get the buy-in from project owners that they will indeed merge smaller sub-tasks, and only then to start working on it.

**Comment by DannyS712** [comment]
@Huji @Amorymeltzer in that case, nevermind

**Comment by siddharthvp** [comment]
I have been thinking, I could take this on if I could get a [Rapid Grant](https://meta.wikimedia.org/wiki/Grants:Project/Rapid) for this. I recently came across [a grant proposal](https://meta.wikimedia.org/wiki/Grants:Project/Rapid/Sophivorus/Excerpts) - which is for improvement to a lua module, and more relatedly [an earlier grant](https://meta.wikimedia.org/wiki/Grants:Project/Rapid/Enhance_the_ProveIt_gadget,_part_2) by the same person for improvement to ProveIt.  As such, i18n of Twinkle too appears to be in scope. 

Are the collaborators okay with this? @Amorymeltzer @MusikAnimal ? 

I did begun the first most preliminary work on localisation with #747 and #758. But have not made further progress (mostly because of morebits-related work).  I have been having a rough plan in mind regarding localisation for some time now. I can draw up a roadmap in a new issue. 

I am yet to fully review the terms and and all, but would like to hear from others first. Seeing that this has remained stalled for 7 years, this does look like a right kind of task to do on payment.

**Comment by Amorymeltzer** [comment]
I looked into it myself, actually, but [item 6](https://meta.wikimedia.org/wiki/Grants:Project/Rapid/Learn) (`Funds are for direct expenses, and may not be used to pay for people’s time.`) seemed to preclude it (also, lol at $2000 paying for childcare); those examples seem to belie that belief, so maybe it would be reasonable?  MA might know more on that.

If it interests you, I don't think I have a problem?  It'd be good to work out a roadmap first (that'd be good regardless, as this issue probably isn't the best place).  I can't promise speedy reviews, but i18n is somewhat amenable to smaller, piecemeal PRs (unlike #414), which are *infinitely* easier to test/review, so it should at least get done on my end.

**Comment by MusikAnimal** [comment]
I don't know how grants work, but regardless, I'm certainly okay with making TW localizable!

We do need a roadmap. It looks like you're starting with making features wiki-agnostic, which is great. I left some thoughts about this at https://phabricator.wikimedia.org/T108426#1545587. The overall goal here should be "don't assume enwiki" and provide only the common logic, and from there each wiki is expected to extend it to suit their needs. I think this may involve some research. We need to see just how different each TW installation is. https://tools.wmflabs.org/global-search/?q=Twinkle&namespaces=8 might be a good starting point.

In the long run, I think Twinkle needs to be made into a "global" gadget, similar to how I did it for MoreMenu. All the source files live on Meta, and gadget definitions on each wiki will source ext.gadget.Twinkle-global (don't call it just Twinkle, because it will clash with local gadget definitions). From there, wikis can extend and customize Twinkle with local JS that's put in the gadget definition.

For translations, I again suggest something similar to [MoreMenu](https://meta.wikimedia.org/wiki/Special:PrefixIndex/MediaWiki:Gadget-MoreMenu.messages) (HotCat has a similar approach), where the English is part of the Meta gadget definition, and we use AJAX to [pull in the appropriate translations](https://github.com/MusikAnimal/MoreMenu/blob/master/src/MoreMenu.js#L168-L190) based on `wgUserLanguage`. The great thing about this is when you browse to another wiki, Twinkle will still be in your language, not the wiki's (except for local extensions to TW that have hard-coded messages). Longer-term, we can use translatewiki.net, and have our own script to generate the message JS files (since JSON can't be put behind ResourceLoader).

Overall, I doubt you'll be able to do all of this within the confines of a grant. You'll also need a lot of help from us/others, and I assume we're not getting paid. Not that I care, I'm not in it for the money :) That said, don't let me stop you from trying.

**Comment by Amorymeltzer** [comment]
>The overall goal here should be "don't assume enwiki" and provide only the common logic

This is probably better left to the afore-described working/tracking issue, but I think far-and-away the hardest issue is going to be dealing with the potential for different processes being handled differently, even down to "does this wiki prepend or append this particular xfd" type stuff.

**Comment by MusikAnimal** [comment]
>  the hardest issue is going to be dealing with the potential for different processes being handled differently

For sure. It's this that makes me wonder what should be translatable, since the processes may very and hence so will the wording. As a proof of concept we might start with the diff module, since it's among the simplest and has universal compatibility. As @huji hints at, baby steps are probably the only way forward.

**Comment by Huji** [comment]
That is half of what I said, @MusikAnimal and you are correct about it. The other half is that those who have contributed to Twinkle significantly and/or for longer and therefore have a sense of "ownership" to its code (or carry a weight as to which direction the code will go) should also show signs that they are willing to loosen their grip on this matter. With just one of the two, this issue will never be resolved.

I rarely take such a hard, political stance of code related stuff. This issue is one of the exceptions. I am not saying what I said above lightly. If I were to make a bet right now, i would say this issue would not even be close to a resolution by Dec 2020. I don't mean to discourage @siddharthvp I just want to let him/her know that the problem here is not just lack of resources (something grants could help with); it is a lack of political will to forgo Twinkle as is, to let a new version of it to arise.

**Comment by siddharthvp** [comment]
Sorry for not following up on this (got busy for a while). It turns out that I don't have 2 months of contiguous free time (which will likely be needed for the work) until later this year. I might try for the grant and undertake this later this year (starting September(?)). 

FWIW, I am aware of all the challenges mentioned above, and plan to deal with them. XFD is a tough nut to crack because of diversity in process across wikis, but let's see ...

**Comment by Huji** [comment]
FWIW, at fawiki we decided to translate and localize a more recent version of Tinkle (the one we officially use is very old and full of patch work). In doing so, we are marking every line of code that we are localizing (see [[ https://fa.wikipedia.org/wiki/%D9%85%D8%AF%DB%8C%D8%A7%D9%88%DB%8C%DA%A9%DB%8C:Gadget-twinklefluff-2020.js | this ]] for instance, and search for `// localized`). By the end of this process, we should have identified most (all?) parts of code that need to be i18n'ed and/or l10n'ed. Obviously, as the source code moves on, this will become less useful. In short: here is a once in a long time opportunity for the developers to start i18n'ing the code.

**Comment by 1ec5** [comment]
I also maintain [a fork](https://github.com/1ec5/twinkle/) of this repository customized for the Vietnamese Wikipedia, which could serve as another data point for internationalization. I generally don’t mark the individual changed lines, but you can [compare the `vi` branch to `master`](https://github.com/1ec5/twinkle/compare/master...vi) for the same effect. There may be similar forks of this repository in other languages.

**Comment by siddharthvp** [comment]
I plan to make a Rapid Grant request this month. @Huji said this won't get done by Dec 2020, so I still have 3 months to prove him wrong ;-) The work would also involve some cleanup of the codebase, upgrade to use ES2020 (using Babel as transpiler), possibly setting up unit tests. Given the size of the codebase, none of that is easy to do without introducing a ton of bugs (as we saw with #908). Even if I'm not entirely successful with the i18n there will still be a lot of improvements done along the way which would be useful for future Twinkle developers (both local and other wikis').

I'd want to have merge access to this repo before I begin though. I wouldn't merge anything non-trivial without seeking code review, but my time is limited so I don't want to have to bother with them for merging small bug fixes and the like arising from previous commits.

**Comment by Huji** [comment]
I would love to be proven wrong.

**Comment by atlight** [comment]
@siddharthvp you'd need to ask @azatoth for contributor rights, I imagine. Perhaps it is time Twinkle was moved to a GitHub organisation?

**Comment by siddharthvp** [comment]
I like the idea of moving to a github organisation. Opened #1140 to discuss that.

**Comment by MusikAnimal** [comment]
I support giving siddharthvp merge access (with or without the grant), but I would recommend the "new" version of Twinkle go in a separate branch until we are convinced it is stable, in the event urgent fixes and such are needed on master.

**Comment by Amorymeltzer** [comment]
>upgrade to use ES2020 (using Babel as transpiler),

Is restructuring the entire codebase going to make it easier for you?  I would think that would delay the goal of the grant.

I had thought to look into this before covid took away any contiguous chunks of time, but is there a reason what @MusikAnimal does for [MoreMenu](https://github.com/MusikAnimal/MoreMenu) couldn't work here?  The infrastructure doesn't look too bad.

**Comment by MusikAnimal** [comment]
I don't think there's an overwhelming need to convert old code to ES2020, but there's little harm in adding a build step with Babel such that any _new_ code can be written in ES2020. I thought @siddharthvp documented how to use a local server to test your changes on the wiki, but I don't see that at [CONTRIBUTING.md](https://github.com/azatoth/twinkle/blob/master/CONTRIBUTING.md). You'll definitely want to go this route moving forward, since it will be too much hassle to build the assets every time you want to test your changes during development. Assuming you're using a modern browser, you just import the untranspiled code, and only run `npm run build` when you're ready to create your PR.

It's still unclear to me what exactly the plan is for internationalizing Twinkle. If you wanted to do this right, with future-proofing in mind, modules like XFD wouldn't go in the "core" Twinkle package, since it is specific to English Wikipedia. You also need to worry about how messages are stored. Further up I advocated for the system I built for MoreMenu, but more ideally we'd use translatewiki.net such that translations are stored in the repo and changes to messaging get translated more quickly. I have a branch to set that up for MoreMenu but I never finished it.

I don't really participate as much here these days, as you know, so don't let me call any shots. But I think it would be beneficial to write out a concrete plan before starting this project.

**Comment by Amorymeltzer** [comment]
@MusikAnimal See #1142 for some documentation on it.

As I see it, there are two related but separate issues.

One — "allow i18n/l10n" — is more straightforward, basically asking to create a system whereby the strings and descriptions and so on can be easily customized, ideally in a single file (see @Huji's comment from June).  I think what MoreMenu does, or something like #329, would accomplish this.

The second is to make it easier for each project wanting to use Twinkle to make it suit their wiki.  IMO this is much more complicated and fraught.  #793 and #946 allow(ed) this in a crude, blunt sort of way, by making it easy to simply turn off things that projects didn't want, but that's obviously minimally helpful.  This would be a truly massive amount of work, even beyond just externalizing lists of templates and such, as so much of the structure in things like xfd, tag, csd, are designed around enwiki's peccadilloes.

**Comment by siddharthvp** [comment]
@MusikAnimal My plan for localisation is to make the modules object-oriented with hierarchical inheritance. For example, the tag module has 3 modes (articles/files/redirects). We want the number of modes (and the conditions in which each one is active) to be customisable, in addition to the list of templates available in each mode. So, I'll have an abstract TagMode class in the "core", and 3 derived classes ArticleMode, RedirectMode and FileMode for enwiki. Things that the modes share in common will be in one of the base classes. [A WIP bare-bones version](https://gist.github.com/siddharthvp/d8e514016c1e3b3c9497866c9f8bb0a9).

Similarly, for XFD, there'll be an abstract XfdProcess class, and Afd, Mfd, Tfd, Ffd, Cfd, Rfd classes extending it. CfdMerge, CfdRename, CfdDelete would then extend Cfd (to account for the differences in these sub-processes), and so on. At the top of the hierarchy, we'd probably want to have a TwinkleModule class to share code between modules.

This would entail individual projects to write portions of code on their own (I don't see a way out of that, because it isn't about how we do it that's different, but it's about _what_ we do is different b/w wikis). To ensure that this will be an easy process, we can't have things like morebits.wiki.page with its unusal interface, pathetic error handling, callback-based layout etc. To address that I'll most probably write a modern Api and Page class that extend the goodness of mw.Api and mw.Title, and which is free of the backward compatibility requirements of morebits.wiki.*

----

I18n of strings isn't something I've thought deeply about. I'm not familiar with translatewiki.net, but I think it is only for per-language translations, not per-project translations, right? While there are some strings which don't vary between projects in the same language (the ones in fluff and diff modules, morebits status messages and the like), these are the minority. The majority of the strings would need to be different for every wiki, and the l10n of the processes I guess does away the need to provide  for i18n of these strings (since the strings are part of the processes).

I'll write a more elaborate summary of the plans and create the grant request page in a few days time.

**Comment by siddharthvp** [comment]
I've put together the rough plan in #1144.

**Comment by siddharthvp** [comment]
Hey folks! I have opened the Rapid Grant request page at [meta:Grants:Project/Rapid/SD0001/Twinkle_localisation](https://meta.wikimedia.org/wiki/Grants:Project/Rapid/SD0001/Twinkle_localisation). If you have any feedback on changing anything on there, please share. Otherwise, I look forward to your [endorsements](https://meta.wikimedia.org/w/index.php?title=Grants:Project/Rapid/SD0001/Twinkle_localisation&action=edit&section=6)!

**Comment by Huji** [comment]
9.5 years pass since I opened the current issue. Aside from all other efforts, a $2000 project was specifically conducted to address this need and yet we don't have an internationalized version of twinkle in use anywhere.

As I read the Learnings of the project at https://meta.wikimedia.org/wiki/Grants:Project/Rapid/SD0001/Twinkle_localisation/Report I come to one conclusion: the only way i18n can happen for Twinkle is by issuing a code freeze on Twinkle--except for security issues--and postponing all enhancements and modifications until the i18n backbone has (a) been completed to a deployable state and (b) has been deployed on English Wikipedia.

I think #1375 is a great idea, and delaying it or deferring to another wiki than English Wikipedia would indicate an unwillingness to let #129 to become a reality.

**Comment by siddharthvp** [comment]
People have started making use of twinkle-starter (see https://github.com/wikimedia-gadgets/twinkle-core/network/dependents?package_id=UGFja2FnZS0xNzQ4MzQwMzYx). There's also some active discussion on the [Discord server](https://discord.gg/P9mqtjBDNb). 

> As in, would it be trivial to get at least a couple of modules working?

Yes twinkle-starter supports Fluff and Diff modules out of the box. The remaining modules require some configuration. You can refer to the [core documentation](https://tools-static.wmflabs.org/twinkle/core-docs/) and [enwiki configurations](https://github.com/wikimedia-gadgets/twinkle-enwiki/blob/master/src). There is a scope for improving twinkle-starter to make it easier to write these configurations (such as by proving some dummy configuration with dummy template names, etc) which others can modify. So please let us know your feedback if you're working on this.
